### PR TITLE
using lookahead in zkd

### PR DESCRIPTION
**Pull request by romanatarango**
### Scope & Purpose

This is an experimental improvement of the ZKD based index iterator. While querying a multidimensional range of values one can now choose (instead of computing the next intersection point of the Z-curve with the range) to try first several documents that are immediately after the current document. If one of them is in the range, the computation of the next intersection point is not necessary, otherwise the point is computed.

A new parameter can be passed from the query: `lookahead` (a non-negative number). It tells how many documents after the current one should be checked for being in the search range before looking for the next point where the Z-curve intersects the search range. The lookahead is passed from the query to, finally, `RocksDBZkdIndexIterator`. However, it possibly should depend on the dimension of the index and on the limits of the query (see function `numNextTries`). The main change is in the function `RocksDBZkdIndexIterator::nextImpl()`.

- [ ] :hankey: Bugfix (requires CHANGELOG entry)
- [ ] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [x] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [ ] :book: CHANGELOG entry made

#### Backports:

- [ ] No backports required
- [ ] Backports required for: *(Please specify versions and link PRs)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket number:
- [ ] Design document: 

### Testing & Verification

*(Please pick either of the following options)*

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [x] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [ ] This PR adds tests that were used to verify all changes:
  - [ ] Added new C++ **Unit tests**
  - [ ] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

Link to Jenkins PR run:

### Documentation

> All new features should be accompanied by corresponding documentation. 
> Bugs and features should furthermore be documented in the CHANGELOG so that
> support, end users, and other developers have a concise overview. 

- [ ] Added entry to *Release Notes* 
- [ ] Added a new section in the *Manual* 
- [ ] Added a new section in the *HTTP API* 
- [ ] Added *Swagger examples* for the HTTP API  
- [ ] Updated license information in *LICENSES-OTHER-COMPONENTS.md* for 3rd party libraries

### CODE REVIEW

File: arangod/Aql/IndexHint.h [hunk 1]
```diff
@@ -56,9 +56,11 @@ class IndexHint {
   std::string typeName() const;
   std::string toString() const;
 
+  size_t getLookahead() const noexcept { return _lookahead; }
  private:
   HintType _type;
   bool _forced;
+  size_t _lookahead = 1;
```
Comment (maierlars) [inline arangod/Aql/IndexHint.h:63]: Here we have default value 1.

File: arangod/Indexes/IndexIterator.h [hunk 2]
```diff
@@ -285,6 +285,9 @@ struct IndexIteratorOptions {
   bool evaluateFCalls = true;
   /// @brief enable caching
   bool enableCache = true;
+  /// @brief number of lookahead elements considered before computing the next
+  /// intersection of the Z-curve with the search range
+  size_t lookahead = 2;
```
Comment (maierlars) [inline arangod/Indexes/IndexIterator.h:290]: And here we have a default value of 2. This can be confusing.

**Review by maierlars — CHANGES_REQUESTED** [review]
There are zdk index js tests. we should run another set of those tests with lookahead enabled to check that it does not miss any documents. See `tests/js/server/aql/aql-optimizer-zkdindex-multi.js`

File: arangod/Indexes/IndexIterator.h (hunk 2 above)
Comment (romanatarango) [reply arangod/Indexes/IndexIterator.h:290]: made 1 as well

File: arangod/Aql/IndexHint.h (hunk 1 above)
Comment (romanatarango) [reply arangod/Aql/IndexHint.h:63]: Now the default is everywhere 1.

File: arangod/Indexes/IndexIterator.h (hunk 2 above)
Comment (romanatarango) [reply arangod/Indexes/IndexIterator.h]: Now the default is everywhere 1.

**Comment by romanatarango** [comment]
> There are zdk index js tests. we should run another set of those tests with lookahead enabled to check that it does not miss any documents. See `tests/js/server/aql/aql-optimizer-zkdindex-multi.js`

Added testing with loolkahead 32.

**Review by maierlars — APPROVED** [review]

**Review by mpoeter — APPROVED** [review]

**Pull request merged**